### PR TITLE
Add castle wars bandages heal amount

### LINKED ISSUE #7862: Castle Wars Bandages heal amount
Currently, the Castle Wars Bandages only shows the run energy recovered. However, the amount they heal would be very welcome information. My proposal would be to add the amount healed (along with the currently displayed run energy) in the overlay when holding your mouse over a Castle Wars Bandage.

Additionally, the amount healed is dependent on the use of a Castle Wars Bracelet and the total hitpoints of the user. Both of these factors would have to be taken into account.

### PR DESCRIPTION
**Pull request by Giovds**
#### Issue
Closes #7862 
Castle Wars bandages only showed run energy when hovering over the item.

#### Add
New class to show both run energy and healing.
Checks if player has Castle Wars bracelet equipped to modify the amount of healing.

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+		return changes;
+	}
+
+	private boolean hasBracelet(ItemContainer equipmentContainer)
+	{
+		if (equipmentContainer != null)
+		{
+			final Item[] equipment = equipmentContainer.getItems();
+			if (equipment.length >= GLOVES_SLOT_ID)
```
Comment (TheStonedTurtle) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: `>=` is wrong here, should be `>`

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+				Stream.of(changes.getStatChanges())
+				.map(sc -> sc.getPositivity())
+				.max(Comparator.comparing(Enum::ordinal)).get());
+
+		return changes;
+	}
+
+	private boolean hasBracelet(ItemContainer equipmentContainer)
+	{
+		if (equipmentContainer != null)
```
Comment (TheStonedTurtle) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: Change to a guard clause?

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	{
+		final StatsChanges changes = new StatsChanges(2);
+		final ItemContainer equipmentContainer = client.getItemContainer(InventoryID.EQUIPMENT);
+		final double percH = hasBracelet(equipmentContainer) ? BRACELET_HP_PERC : BASE_HP_PERC;
+		final SimpleStatBoost hitPoints = heal(HITPOINTS, perc(percH, 0));
+
+		changes.setStatChanges(
+				Stream.concat(
+				Stream.of(hitPoints.effect(client)),
+				Stream.of(RUN_ENERGY_BOOST.effect(client))
```
Comment (TheStonedTurtle) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: Wouldn't `Stream.of(hitPoints.effect(client), RUN_ENERGY_BOOST.effect(client))` work instead? Is there a reason you are creating two separate streams then combining them?

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+import java.util.Comparator;
+import java.util.stream.Stream;
+
+import static net.runelite.client.plugins.itemstats.Builders.heal;
+import static net.runelite.client.plugins.itemstats.Builders.perc;
+import static net.runelite.client.plugins.itemstats.stats.Stats.HITPOINTS;
+import static net.runelite.client.plugins.itemstats.stats.Stats.RUN_ENERGY;
+
+@RequiredArgsConstructor
```
Comment (TheStonedTurtle) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: This seems unused?

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import net.runelite.client.plugins.itemstats.StatChange;
+import net.runelite.client.plugins.itemstats.StatsChanges;
+
+import java.util.Comparator;
+import java.util.stream.Stream;
+
+import static net.runelite.client.plugins.itemstats.Builders.heal;
+import static net.runelite.client.plugins.itemstats.Builders.perc;
+import static net.runelite.client.plugins.itemstats.stats.Stats.HITPOINTS;
+import static net.runelite.client.plugins.itemstats.stats.Stats.RUN_ENERGY;
```
Comment (TheStonedTurtle) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: Fix your import order, you should be able to auto fix with `ctrl+alt+o` (Code->Optimize Imports).

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/ItemStatChanges.java [hunk 6]
```diff
@@ -137,6 +138,9 @@ private void init()
 		// Regular overload (NMZ)
 		add(combo(5, boost(ATTACK, perc(.15, 5)), boost(STRENGTH, perc(.15, 5)), boost(DEFENCE, perc(.15, 5)), boost(RANGED, perc(.15, 5)), boost(MAGIC, perc(.15, 5)), heal(HITPOINTS, -50)), OVERLOAD_1, OVERLOAD_2, OVERLOAD_3, OVERLOAD_4);
 
+		// Bandages (Castle Wars)
+		add(new CasteWarsBandage(), RAW_LOBSTER);
```
Comment (TheStonedTurtle) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemstats/ItemStatChanges.java]: RAW_LOBSTER?

Comment (Giovds) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemstats/ItemStatChanges.java]: My bad, forgot to remove test data

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java (hunk 2 above)
Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: I think this is always non null anyway.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+public class CasteWarsBandage implements Effect
+{
+	private static final ImmutableSet<Integer> BRACELETS = ImmutableSet.of(
+			ItemID.CASTLE_WARS_BRACELET1, ItemID.CASTLE_WARS_BRACELET2, ItemID.CASTLE_WARS_BRACELET3
+	);
+	private static final int GLOVES_SLOT_ID = EquipmentInventorySlot.GLOVES.getSlotIdx();
+	private static final double BASE_HP_PERC = .10;
+	private static final double BRACELET_HP_PERC = .50;
+	private static final StatsChanges CHANGES = new StatsChanges(2);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: This should probably just be a local variable?

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java (hunk 2 above)
Comment (TheStonedTurtle) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: I'll have to double check, was pretty sure it was null If you aren't wearing anything when you login.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java (hunk 7 above)
Comment (Giovds) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: I was told it was preferred to use as many global variables as possible. Are locals preferred?

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2016-2018, Adam <Adam@sigterm.info>
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: You should probably fix your copyright header, unless you are Adam

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+import net.runelite.client.plugins.itemstats.StatsChanges;
+import static net.runelite.client.plugins.itemstats.stats.Stats.HITPOINTS;
+import static net.runelite.client.plugins.itemstats.stats.Stats.RUN_ENERGY;
+
+public class CasteWarsBandage implements Effect
+{
+	private static final ImmutableSet<Integer> BRACELETS = ImmutableSet.of(
+			ItemID.CASTLE_WARS_BRACELET1, ItemID.CASTLE_WARS_BRACELET2, ItemID.CASTLE_WARS_BRACELET3
+	);
+	private static final int GLOVES_SLOT_ID = EquipmentInventorySlot.GLOVES.getSlotIdx();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: Add empty line above

Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: Also I dont think this variable is even necessary?

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java (hunk 7 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: If you are mutating static final field you are usually doing something wrong

File: runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+	{
+		if (equipmentContainer != null)
+		{
+			final Item[] equipment = equipmentContainer.getItems();
+			if (equipment.length > GLOVES_SLOT_ID)
+			{
+				return BRACELETS.contains(equipment[GLOVES_SLOT_ID].getId());
+			}
+		}
+		return false;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemstats/special/CasteWarsBandage.java]: Add empty line above

**Pull request merged**